**Анализ данных о бронировании отелей**

В первом мини-проекте вам предстоит проанализировать данные о бронировании отелей. 

Hotel – тип отеля (City Hotel или Resort Hotel) 

Is canceled – бронирование было отменено (1) или нет (0); не отмененное считается успешным

Lead time – количество дней, прошедших между датой бронирования и датой прибытия  

Arrival full date – полная дата прибытия

Arrival date year – год прибытия  

Arrival date month – месяц прибытия  

Arrival date week number – номер недели прибытия

Arrival date day of month – день прибытия

Stays in weekend nights – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в отеле

Stays in week nights – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле

Stays total nights – общее число забронированных ночей (сумма двух предыдущих колонок)

Adults – число взрослых

Children – число детей

Babies – число младенцев 

Meal – выбранный тип питания

Country – страна происхождения клиента

Reserved room type – тип зарезервированного номера

Assigned room type – тип полученного номера (может отличаться от забронированного)

Customer type – тип бронирования

Reservation status – значение последнего статуса брони: Canceled - было отменено клиентом; Check-Out - клиент
зарегистрировался, но уже покинул отель; No-Show - клиент не зарегистрировался и сообщил администрации отеля причину

Reservation status date – дата обновления статуса

Задания:

- Импортируйте библиотеку pandas как pd. Загрузите датасет bookings.csv с разделителем ;. Проверьте размер таблицы, типы переменных, а затем выведите первые 7 строк, чтобы посмотреть на данные. 


- Приведите названия колонок к нижнему регистру и замените пробелы на знак нижнего подчеркивания.


- Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.


- На сколько ночей в среднем бронируют отели разных типов?


- Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?


- Проанализируйте даты запланированного прибытия. 
– На какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли самый популярный месяц в 2017?
– Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в каждый из периодов


- Посмотрите на числовые характеристики трёх переменных: adults, children и babies. Какая из них имеет наибольшее среднее значение?


- Создайте колонку total_kids, объединив children и babies. Отели какого типа в среднем пользуются большей популярностью у клиентов с детьми?


- Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Посчитайте отношение количества ушедших пользователей к общему количеству клиентов, выраженное в процентах (churn rate). Укажите, среди какой группы показатель выше.

In [1]:
import pandas as pd


In [2]:
df = pd.read_csv('bookings (1).csv', sep=';')

In [3]:
df

Hotel  Is Canceled  Lead Time arrival full date  \
0       Resort Hotel            0        342        2015-07-01   
1       Resort Hotel            0        737        2015-07-01   
2       Resort Hotel            0          7        2015-07-01   
3       Resort Hotel            0         13        2015-07-01   
4       Resort Hotel            0         14        2015-07-01   
...              ...          ...        ...               ...   
119385    City Hotel            0         23        2017-08-30   
119386    City Hotel            0        102        2017-08-31   
119387    City Hotel            0         34        2017-08-31   
119388    City Hotel            0        109        2017-08-31   
119389    City Hotel            0        205        2017-08-29   

        Arrival Date Year Arrival Date Month  Arrival Date Week Number  \
0                    2015               July                        27   
1                    2015               July                        27   
2                    2015               July                        27   
3                    2015               July                        27   
4                    2015               July                        27   
...                   ...                ...                       ...   
119385               2017             August                        35   
119386               2017             August                        35   
119387               2017             August                        35   
119388               2017             August                        35   
119389               2017             August                        35   

        Arrival Date Day of Month  Stays in Weekend nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        Stays in week nights  ...  Adults  Children  Babies  Meal Country  \
0                          0  ...       2       0.0       0    BB     PRT   
1                          0  ...       2       0.0       0    BB     PRT   
2                          1  ...       1       0.0       0    BB     GBR   
3                          1  ...       1       0.0       0    BB     GBR   
4                          2  ...       2       0.0       0    BB     GBR   
...                      ...  ...     ...       ...     ...   ...     ...   
119385                     5  ...       2       0.0       0    BB     BEL   
119386                     5  ...       3       0.0       0    BB     FRA   
119387                     5  ...       2       0.0       0    BB     DEU   
119388                     5  ...       2       0.0       0    BB     GBR   
119389                     7  ...       2       0.0       0    HB     DEU   

       Reserved Room Type Assigned room type customer type Reservation Status  \
0                       C                  C     Transient          Check-Out   
1                       C                  C     Transient          Check-Out   
2                       A                  C     Transient          Check-Out   
3                       A                  A     Transient          Check-Out   
4                       A                  A     Transient          Check-Out   
...                   ...                ...           ...                ...   
119385                  A                  A     Transient          Check-Out   
119386         

In [4]:
#размер датафрейма
print(f' Количество строк: {df.shape[0]}\n Количество колонок: {df.shape[1]}')

 Количество строк: 119390
 Количество колонок: 21


In [5]:
#посмотрим на типы данных, преобладающих в таблице
df.dtypes
#большую часть занимают численные признаки

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

In [6]:
df.dtypes.value_counts()

object     10
int64      10
float64     1
dtype: int64

In [7]:
#название всех необработанных названий колонок в списке
names_columns = df.columns

In [8]:
#функция, которая приводит названия колонок в стандартный вид и возвращает готовый словарь для переименования 
def update_name_columns(df_columns):
    update_names = []
    result_dict = {}
    for i in df_columns:
        new_i = ("_".join(i.split(" "))).lower()
        update_names.append(new_i)
    for j in range(len(df_columns)):
        result_dict[df_columns[j]] = update_names[j]
    return result_dict

In [ ]:
'''
или проще
def to_lower_underscore(name):
    name = name.lower.replace(' ','_')
    return name
    
df = df.rename(columns=to_lower_underscore)
'''

In [9]:
#результирующий словарь
result = update_name_columns(names_columns)
result

{'Hotel': 'hotel',
 'Is Canceled': 'is_canceled',
 'Lead Time': 'lead_time',
 'arrival full date': 'arrival_full_date',
 'Arrival Date Year': 'arrival_date_year',
 'Arrival Date Month': 'arrival_date_month',
 'Arrival Date Week Number': 'arrival_date_week_number',
 'Arrival Date Day of Month': 'arrival_date_day_of_month',
 'Stays in Weekend nights': 'stays_in_weekend_nights',
 'Stays in week nights': 'stays_in_week_nights',
 'stays total nights': 'stays_total_nights',
 'Adults': 'adults',
 'Children': 'children',
 'Babies': 'babies',
 'Meal': 'meal',
 'Country': 'country',
 'Reserved Room Type': 'reserved_room_type',
 'Assigned room type': 'assigned_room_type',
 'customer type': 'customer_type',
 'Reservation Status': 'reservation_status',
 'Reservation status_date': 'reservation_status_date'}

In [10]:
df = df.rename(columns=result)

In [11]:
#обновленный датафрейм с измененными названиями
df

hotel  is_canceled  lead_time arrival_full_date  \
0       Resort Hotel            0        342        2015-07-01   
1       Resort Hotel            0        737        2015-07-01   
2       Resort Hotel            0          7        2015-07-01   
3       Resort Hotel            0         13        2015-07-01   
4       Resort Hotel            0         14        2015-07-01   
...              ...          ...        ...               ...   
119385    City Hotel            0         23        2017-08-30   
119386    City Hotel            0        102        2017-08-31   
119387    City Hotel            0         34        2017-08-31   
119388    City Hotel            0        109        2017-08-31   
119389    City Hotel            0        205        2017-08-29   

        arrival_date_year arrival_date_month  arrival_date_week_number  \
0                    2015               July                        27   
1                    2015               July                        27   
2                    2015               July                        27   
3                    2015               July                        27   
4                    2015               July                        27   
...                   ...                ...                       ...   
119385               2017             August                        35   
119386               2017             August                        35   
119387               2017             August                        35   
119388               2017             August                        35   
119389               2017             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  ...  adults  children  babies  meal country  \
0                          0  ...       2       0.0       0    BB     PRT   
1                          0  ...       2       0.0       0    BB     PRT   
2                          1  ...       1       0.0       0    BB     GBR   
3                          1  ...       1       0.0       0    BB     GBR   
4                          2  ...       2       0.0       0    BB     GBR   
...                      ...  ...     ...       ...     ...   ...     ...   
119385                     5  ...       2       0.0       0    BB     BEL   
119386                     5  ...       3       0.0       0    BB     FRA   
119387                     5  ...       2       0.0       0    BB     DEU   
119388                     5  ...       2       0.0       0    BB     GBR   
119389                     7  ...       2       0.0       0    HB     DEU   

       reserved_room_type assigned_room_type customer_type reservation_status  \
0                       C                  C     Transient          Check-Out   
1                       C                  C     Transient          Check-Out   
2                       A                  C     Transient          Check-Out   
3                       A                  A     Transient          Check-Out   
4                       A                  A     Transient          Check-Out   
...                   ...                ...           ...                ...   
119385                  A                  A     Transient          Check-Out   
119386         

- Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.

In [12]:
is_canceled_countries = df \
.query('is_canceled == 0') \
.groupby('country',as_index=False) \
.agg({'is_canceled':'value_counts'}) \
.sort_values('is_canceled', ascending=False)

In [13]:
is_canceled_countries

country  is_canceled
125     PRT        21071
57      GBR         9676
54      FRA         8481
50      ESP         6391
42      DEU         6069
..      ...          ...
138     SMR            1
101     MLI            1
136     SLE            1
94      MAC            1
122     PLW            1

[165 rows x 2 columns]

In [14]:
#топ - 5 
is_canceled_countries.head(5)

country  is_canceled
125     PRT        21071
57      GBR         9676
54      FRA         8481
50      ESP         6391
42      DEU         6069

Таким образом, пользователи из таких  стран  как (топ 5): PRT, GBR, FRA, ESP, DEU совершили наибольшее число успешных бронирований.

- На сколько ночей в среднем бронируют отели разных типов?

In [20]:
#stays_total_nights, hotel
stays_total_nights_and_hotels = df \
.groupby('hotel', as_index=False) \
.agg({'stays_total_nights':'mean'}) \
.round(decimals=2)

In [21]:
stays_total_nights_and_hotels

hotel  stays_total_nights
0    City Hotel                2.98
1  Resort Hotel                4.32

В отели типа City Hotel	в среднем бронируют на 3 ночи, а в отели Resort Hotel в среднем на 4 ночи.

- Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?

In [16]:
(df.assigned_room_type != df.reserved_room_type).value_counts()

False    104473
True      14917
dtype: int64

In [17]:
(df.assigned_room_type != df.reserved_room_type).sum()

14917

In [ ]:
'''
иначе можно сделать так

df.query('assigned_room_type != reserved_room_type').shape[0]
'''

Таким образом, в датасете встретилось около 14917 случаев, когда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). 

- Проанализируйте даты запланированного прибытия. – На какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли самый популярный месяц в 2017? – Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в каждый из периодов

In [25]:
month_is_canceled_2016 = df \
.query('is_canceled == 0') \
.query('arrival_date_year == 2016') \
.groupby('arrival_date_month', as_index=False) \
.agg({'is_canceled':'value_counts'}) \
.sort_values('is_canceled', ascending=False)

In [27]:
month_is_canceled_2016

arrival_date_month  is_canceled
10            October         3689
8                 May         3563
11          September         3372
0               April         3367
7               March         3347
1              August         3238
6                June         3196
5                July         3073
9            November         2818
3            February         2554
2            December         2462
4             January         1691

Чаще всего успешно оформляли бронь ( 2016 год) в октябре (кол-во успешных бронирований 3689).

In [28]:
month_is_canceled_2017 = df \
.query('is_canceled == 0') \
.query('arrival_date_year == 2017') \
.groupby('arrival_date_month', as_index=False) \
.agg({'is_canceled':'value_counts'}) \
.sort_values('is_canceled', ascending=False)

In [29]:
month_is_canceled_2017

arrival_date_month  is_canceled
7                May         3551
4               July         3329
6              March         3298
5               June         3208
0              April         3198
1             August         3109
2           February         2818
3            January         2431

Чаще всего успешно оформляли бронь ( 2017 год) в мае (кол-во успешных бронирований 3551).

Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в каждый из периодов

In [22]:
city_hotel_month = df \
.query('hotel == "City Hotel" and is_canceled == 1') \
.groupby('arrival_date_year', as_index=False) \
.arrival_date_month \
.value_counts() 

In [23]:
city_hotel_month

arrival_date_year arrival_date_month  count
0                2015          September   1543
1                2015            October   1321
2                2015             August   1232
3                2015               July    939
4                2015           December    668
5                2015           November    301
6                2016            October   1947
7                2016               June   1720
8                2016          September   1567
9                2016              April   1539
10               2016                May   1436
11               2016           November   1360
12               2016             August   1247
13               2016              March   1108
14               2016           December   1072
15               2016               July   1043
16               2016           February    930
17               2016            January    438
18               2017                May   2217
19               2017              April   1926
20               2017               June   1808
21               2017               July   1324
22               2017              March   1278
23               2017             August   1123
24               2017            January   1044
25               2017           February    971

- Посмотрите на числовые характеристики трёх переменных: adults, children и babies. Какая из них имеет наибольшее среднее значение?

In [25]:
mean_adults = df.adults.mean()
mean_children = df.children.mean()
mean_babies = df.babies.mean()
print(f'Среднее значение adults: {mean_adults}\nСреднее значение children: {mean_children}\nСреднее значение babies: {mean_babies}\n')

Среднее значение adults: 1.8564033838679956
Среднее значение children: 0.10388990333874994
Среднее значение babies: 0.007948739425412514



In [24]:
df[['adults', 'children', 'babies']].mean()

adults      1.856403
children    0.103890
babies      0.007949
dtype: float64

Наибольшее среднее значение имеет переменная adults.

- Создайте колонку total_kids, объединив children и babies. Отели какого типа в среднем пользуются большей популярностью у клиентов с детьми?

In [26]:
df['total_kids'] = df['children'] + df['babies']

In [27]:
#основные числовые характеристики колонки
df.children.describe()

count    119386.000000
mean          0.103890
std           0.398561
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max          10.000000
Name: children, dtype: float64

In [28]:
df

hotel  is_canceled  lead_time arrival_full_date  \
0       Resort Hotel            0        342        2015-07-01   
1       Resort Hotel            0        737        2015-07-01   
2       Resort Hotel            0          7        2015-07-01   
3       Resort Hotel            0         13        2015-07-01   
4       Resort Hotel            0         14        2015-07-01   
...              ...          ...        ...               ...   
119385    City Hotel            0         23        2017-08-30   
119386    City Hotel            0        102        2017-08-31   
119387    City Hotel            0         34        2017-08-31   
119388    City Hotel            0        109        2017-08-31   
119389    City Hotel            0        205        2017-08-29   

        arrival_date_year arrival_date_month  arrival_date_week_number  \
0                    2015               July                        27   
1                    2015               July                        27   
2                    2015               July                        27   
3                    2015               July                        27   
4                    2015               July                        27   
...                   ...                ...                       ...   
119385               2017             August                        35   
119386               2017             August                        35   
119387               2017             August                        35   
119388               2017             August                        35   
119389               2017             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  ...  children  babies  meal  country  \
0                          0  ...       0.0       0    BB      PRT   
1                          0  ...       0.0       0    BB      PRT   
2                          1  ...       0.0       0    BB      GBR   
3                          1  ...       0.0       0    BB      GBR   
4                          2  ...       0.0       0    BB      GBR   
...                      ...  ...       ...     ...   ...      ...   
119385                     5  ...       0.0       0    BB      BEL   
119386                     5  ...       0.0       0    BB      FRA   
119387                     5  ...       0.0       0    BB      DEU   
119388                     5  ...       0.0       0    BB      GBR   
119389                     7  ...       0.0       0    HB      DEU   

       reserved_room_type assigned_room_type customer_type reservation_status  \
0                       C                  C     Transient          Check-Out   
1                       C                  C     Transient          Check-Out   
2                       A                  C     Transient          Check-Out   
3                       A                  A     Transient          Check-Out   
4                       A                  A     Transient          Check-Out   
...                   ...                ...           ...                ...   
119385                  A                  A     Transient          Check-Out   
119386                  E                  E     Transient          Check-Out   
119387            

In [29]:
type_hotels_and_childrens_mean = df \
.groupby('hotel', as_index=False) \
.agg({'total_kids':'mean'}) \
.round(decimals=2) 

In [30]:
type_hotels_and_childrens_mean

hotel  total_kids
0    City Hotel        0.10
1  Resort Hotel        0.14

In [31]:
type_hotels_and_childrens_mean.max()

hotel         Resort Hotel
total_kids            0.14
dtype: object

Отели типа Resort Hotel в среднем пользуются большей популярностью у клиентов с детьми.

- Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Посчитайте отношение количества ушедших пользователей к общему количеству клиентов, выраженное в процентах (churn rate). Укажите, среди какой группы показатель выше.

In [32]:
df['has_kids'] = df.total_kids > 0

In [33]:
df.has_kids.value_counts()

False    110058
True       9332
Name: has_kids, dtype: int64

In [34]:
departed_users_without_kids = df \
.query('is_canceled == 1 and has_kids == 0') \
.shape[0]

departed_users_without_kids

40965

In [35]:
all_users_without_kids = df.query('has_kids == 0').shape[0]

In [36]:
churn_rate_without_kids = (departed_users_without_kids / all_users_without_kids)*100

In [37]:
print(f'Отношение количества ушедших пользователей (без детей) к общему количеству клиентов (без детей): {round(churn_rate_without_kids)}%')

Отношение количества ушедших пользователей (без детей) к общему количеству клиентов (без детей): 37%


In [38]:
departed_users_with_kids = df \
.query('is_canceled == 1 and has_kids > 0') \
.shape[0]

departed_users_with_kids

3259

In [39]:
all_users_with_kids = df.query('has_kids > 0').shape[0]
churn_rate_kids = (departed_users_with_kids / all_users_with_kids)*100
print(f'Отношение количества ушедших пользователей (есть дети) к общему количеству клиентов (есть дети): {round(churn_rate_kids)}%')

Отношение количества ушедших пользователей (есть дети) к общему количеству клиентов (есть дети): 35%


Таким образом, churn rate выше у группы пользователей без детей.